src/primitives/transaction.h

```diff
@@ -176,6 +214,48 @@ class CTxOut
     std::string ToString() const;
 };
 
+class CTransaction;
+
+/// <summary>
+/// A generic transaction output that could either be an MWEB output commitment or a canonical CTxOut.
+/// </summary>
+class CTxOutput
+{
+public:
+    CTxOutput() = default;
+    CTxOutput(const CTransaction* pTx, Commitment commitment)
+        : m_tx(pTx), m_idx(commitment), m_output(std::move(commitment)) {}
+    CTxOutput(const CTransaction* pTx, OutputIndex idx, CTxOut txout)
+        : m_tx(pTx), m_idx(std::move(idx)), m_output(std::move(txout)) {}
+
+    bool IsMWEB() const noexcept { return m_output.type() == typeid(Commitment); }
+
+    const OutputIndex& GetIndex() const noexcept { return m_idx; }
+    
+    const Commitment& GetCommitment() const noexcept
+    {
+        assert(IsMWEB());
+        return boost::get<Commitment>(m_output);
+    }
+
+    const CTxOut& GetTxOut() const noexcept
+    {
+        assert(!IsMWEB());
+        return boost::get<CTxOut>(m_output);
+    }
+
+    const CScript& GetScriptPubKey() const noexcept
+    {
+        assert(!IsMWEB());
+        return GetTxOut().scriptPubKey;
+    }
+
+private:
+    const CTransaction* m_tx;
+    OutputIndex m_idx;
+    boost::variant<CTxOut, Commitment> m_output;
```

Comment:
Note that for MWEB we are storing the commitment twice.

Reply:
Hmm, good point. I'll change that.